README.md
[![Go Report Card](https://goreportcard.com/badge/github.com/cmaster11/overseer)](https://goreportcard.com/report/github.com/cmaster11/overseer)
[![license](https://img.shields.io/github/license/cmaster11/overseer.svg)](https://github.com/cmaster11/overseer/blob/master/LICENSE)

Changelog is [here](CHANGELOG.md).

# DISCLAIMER

This project is a heavily modified version of the amazing [skx/overseer](https://github.com/skx/overseer) one. 
Compatibility between the two projects's data is not guaranteed, and should not be expected.

Table of Contents
=================

* [Overseer](#overseer)
* [Installation](#installation)
  * [Kubernetes](#kubernetes)
  * [Dependencies](#dependencies)
* [Executing Tests](#executing-tests)
  * [Parallel execution](#parallel-execution)
  * [Period-tests](#period-tests)
  * [Local testing](#local-testing)
  * [Running Automatically](#running-automatically)
  * [Smoothing Test Failures](#smoothing-test-failures)
* [Notifications](#notifications)
  * [Deduplication](#deduplication)
* [Metrics](#metrics)
* [Redis Specifics](#redis-specifics)

# Overseer

Overseer is a simple and scalable [golang](https://golang.org/)-based remote protocol tester, which allows you to monitor the state of your network, and the services running upon it.

"Remote Protocol Tester" sounds a little vague, so to be more concrete this application lets you test that (remote) services are running, and has built-in support for performing testing against:

* DNS-servers
   * Test lookups of A, AAAA, MX, NS, and TXT records.
* Finger
* FTP
* HTTP & HTTPS fetches.
   * HTTP basic-authentication is supported.
   * Requests may be DELETE, GET, HEAD, POST, PATCH, POST, & etc.
   * SSL certificate validation and expiration warnings are supported.
* IMAP & IMAPS
* Kubernetes service endpoints check
* MySQL
* NNTP
* ping / ping6
* POP3 & POP3S
* Postgres
* redis
* rsync
* SMTP
* SSH
* SSL
* Telnet
* VNC
* XMPP

(The implementation of the protocol-handlers can be found beneath the top-level [protocols/](protocols/) directory in this repository.)

Tests to be executed are defined in a simple text-based format which has the general form:

     $TARGET must run $SERVICE [with $OPTION_NAME $VALUE] ..

You can see what the available tests look like in [the sample test-file](input.txt), and each of the included protocol-handlers are self-documenting which means you can view example usage via:

     ~$ overseer examples [pattern]

All protocol-tests transparently support testing IPv4 and IPv6 targets, although you may globally disable either address family if you wish.

## Installation

To install locally the project:

    git clone https://github.com/cmaster11/overseer
    cd overseer
    go install

### Kubernetes

A sample deployment is provided in the [`example-kubernetes`](./example-kubernetes/) folder. Please take a look at the 
[`README`](./example-kubernetes/README.md) for more instructions.

### Dependencies

Beyond the compile-time dependencies overseer requires a [redis](https://redis.io/) server which is used for two things:

* As the storage-queue for parsed-jobs.
* As the storage-queue for test-results.

Because overseer is executed in a distributed fashion tests are not executed
as they are parsed/read, instead they are inserted into a redis-queue. A worker,
or number of workers, poll the queue fetching & executing jobs as they become
available.

In small-scale deployments it is probably sufficient to have a single worker,
and all the software running upon a single host. For a larger number of
tests (1000+) it might make more sense to have a pool of hosts each running
a worker.

Because we don't want to be tied to a specific notification-system results
of each test are also posted to the same redis-host, which allows results to be retrieved and transmitted to your preferred notifier.

More details about [notifications](#notifications) are available later in this document.

## Executing Tests

As mentioned already executing tests a two-step process:

* First of all tests are parsed and inserted into a redis-based queue.
* Secondly the tests are pulled from that queue and executed.

This might seem a little convoluted, however it is a great design if you
have a lot of tests to be executed, because it allows you to deploy multiple
workers. Instead of having a single host executing all the tests you can
can have 10 hosts, each watching the same redis-queue pulling jobs, & executing
them as they become available.

In short using a central queue allows you to scale out the testing horizontally.

To add your tests to the queue you should run:

    $ overseer enqueue \
        -redis-host=queue.example.com:6379 [-redis-pass='secret.here'] \
        test.file.1 test.file.2 .. test.file.N

This will parse the tests contained in the specified files, adding each of them to the (shared) redis queue. 
Once all of the jobs have been parsed and inserted into the queue the process will terminate.

To drain the queue you can should now start a worker, which will fetch the tests and process them:

    $ overseer worker -verbose \
        -redis-host=queue.example.com:6379 [-redis-pass='secret']

The worker will run constantly, not terminating unless manually killed. With
the worker running you can add more jobs by re-running the `overseer enqueue`
command.

To run tests in parallel simply launch more instances of the worker, on the same host, or on different hosts.

### Parallel execution

By default the worker will process in parallel a number of tests equal to the number of the current machine's logical
CPUs. To alter this behavior, you can use the `-parallel` flag:

    $ # Runs 9 tests at a time
    $ overseer worker -parallel 9
    
Using a higher number of parallel tests is useful if running any long-running tests, to not delay executions of any others.

### Period-tests

Let's imagine that you want to test how many times your web service fails in 1 minute. You can run period-tests:

    https://example.com must run http with pt-duration 60s with pt-sleep 2s with pt-threshold 15%
    
The previous line will trigger a period-test, where the same test `https://example.com must run http` 
will be tested over and over for a duration of 60 seconds (`pt-duration 60s`), with a pause of 2 seconds (`pt-sleep 2s`)
between each test. At the end of the testing period, if the percentage of errors is higher than 15% (`pt-threshold 15%`), 
an alert will be generated, e.g:

    8 tests failed out of 21 (38.10%)
    
You can also test multiple cases with a dumb test:

    dumb-test1 must run dumb-test with pt-duration 5s with pt-sleep 200ms with pt-threshold 0% with dumb-duration-max 100ms
    dumb-test2 must run dumb-test with pt-duration 5s with pt-sleep 200ms with pt-threshold 20% with dumb-duration-max 100ms
    dumb-test3 must run dumb-test with pt-duration 5s with pt-sleep 200ms with pt-threshold 40% with dumb-duration-max 100ms
    
If no `pt-sleep` is defined, Overseer will default to the `-period-test-sleep` command line variable value, or to `5s`.
If no `pt-threshold` is defined, Overseer will default to the `-period-test-threshold` command line variable value, or to `0%`.

Note: the `pt-` flags are shortened versions of the also usable longer tags:
 
    pt-duration -> period-test-duration
    pt-sleep -> period-test-sleep
    pt-threshold -> period-test-threshold
    
Note: period-tests, by default, have no enabled [deduplication](#deduplication) rules. To enable deduplication, you need
to manually add the `with dedup 5m` flag.
    
### Local testing

You can test Overseer functionalities locally using some scripts.

Setup Overseer with:

* Run a local redis: `./scripts/test-run-redis.sh` (hosts the processing queue)
* Run a local worker: `./scripts/test-run-worker.sh` (runs the actual tests)
* Run a local webhook bridge: `./scripts/test-run-webhook-bridge.sh` (fetches test results from the queue and triggers the webhook)
* Run a local http webhook listener: `./scripts/test-run-http-dump.sh` (listens for webhooks requests and dumps them to `stdout`)

Run tests with:

* Sample dumb tests: `./scripts/test-run-enqueue.sh`
* An always-failing test: `./scripts/test-run-enqueue-fail.sh`
* An sample period-test: `./scripts/test-run-enqueue-period.sh`
* Custom rules: `./scripts/test-run-enqueue-stdin.sh "https://google.com must run http"`

### Running Automatically

Beneath [systemd/](systemd/) you will find some sample service-files which can be used to deploy overseer upon a single host:

* A service to start a single worker, fetching jobs from a redis server.
  * The redis-server is assumed to be running on `localhost`.
* A service & timer to regularly populate the queue with fresh jobs to be executed.
  * i.e. The first service is the worker, this second one feeds the worker.

### Smoothing Test Failures

To avoid triggering false alerts due to transient (network/host) failures
tests which fail are retried several times before triggering a notification.

This _smoothing_ is designed to avoid raising an alert, which then clears
upon the next overseer run, but the downside is that flapping services might
not necessarily become visible.

If you're absolutely certain that your connectivity is good, and that
alerts should always be raised for failing services you can disable this
retry-logic via the command-line flag `-retry=false`.

## Notifications

The result of each test is submitted to the central redis-host, from where it can be pulled and used to notify a human of a problem.

Sample result-processors are [included](bridges/) in this repository which post test-results via [webhook](bridges/webhook-bridge/main.go) (e.g. to trigger notifications with [Notify17](https://notify17.net), to a [purppura instance](https://github.com/skx/purppura) or via email.

The sample bridges are primarily included for demonstration purposes, the
expectation is you'll prefer to process the results and issue notifications to
humans via your favourite in-house tool - be it [Notify17](https://notify17.net), or something similar.

The results themselves are published as JSON objects to the `overseer.results` set. Your notifier should remove the results from this set, as it generates alerts to prevent it from growing indefinitely.

You can check the size of the results set at any time via `redis-cli` like so:

    $ redis-cli llen overseer.results
    (integer) 0

The JSON object used to describe each test-result has the following fields:

| Field Name | Field Value                                                                                              |

CHANGELOG.md
# Changelog

## [2020/05/30] cmaster11/overseer:1.13.3

* Overseer now supports some new test options:
    * `min-duration`: when defined, an alert will NOT be triggered until the test error is repeated for AT LEAST `min-duration` time.
        
        Example: `https://google.com must run http with min-duration 1m` will not generate alerts until the same test fails for more than a minute.
        
    * `test-label`: when defined, offers more flexibility to test result messages, where the field `testLabel` is now passed as it is. 
        This option is useful to have "clean" test result messages, where the test label can be shown instead of the original test input.
    
        Example: `https://google.com must run http with test-label "Google test"`

## [2020/05/11] cmaster11/overseer:1.12.12

* Overseer now supports two new runtime period-test variables for HTTP test:

    * `__pt-index__`: current period test index
    * `__pt-time-ms__`: current period test start time
    
    Example: `https://google.com?testIdx=__pt-index__ must run http with connect-timeout 2.5s with connect-retries 2 with pt-duration 1m with pt-sleep 1s with pt-threshold 0%`

## [2020/05/11] cmaster11/overseer:1.12.11

* Sort targets alphabetically when using `max-targets`, which increases the chance of pointing to the same target
    on subsequent runs.

## [2020/05/06] cmaster11/overseer:1.12.10

* HTTP test has new option:
    * `follow-redirect true|(\d+)`: allows following HTTP redirects.
        * `follow-redirect true`: allows max 10 redirects
        * `follow-redirect 25`: allows max 25 (or any other used number) redirects
* All tests which resolve hostnames have a new option:
    * `max-targets 2`: forces the test to run against only the first 2 (or any other number) found targets of the hostname, instead of all of them.

## [2019/12/17] cmaster11/overseer:1.12.8

* BREAKING: new behavior for `k8s-event-watcher` rules. Events will now be marked as errors **only** if `errorRules` rules exist and are matched:
```yaml
filters:
- rules:
    involvedObject.kind: Job
    involvedObject.name: "^*.fail"
    reason: BackoffLimitExceeded
  errorRules:
    # Any matched event is an error
    type: .*
```

## [2019/12/17] cmaster11/overseer:1.12.7

* BREAKING: new version of [k8s-event-watcher](https://github.com/cmaster11/k8s-event-watcher):
    * Filters configuration now accepts any field name, and a new `rules` keyword has been introduced:
    ```yaml
    filters:
    - rules:
        involvedObject.kind: Job
        involvedObject.name: "^*.fail"
        reason: BackoffLimitExceeded
    ```
* You can now use the `connect-retries` flag for HTTP tests, which will have Overseer retry the connection to a server in case of connect timeout.

## [2019/12/16] cmaster11/overseer:1.12.5

* You can now use a per-test `timeout` flag, which will override the worker one if set.
* HTTP test has new timeouts:
    * `connect-timeout`: fails if worker takes too long to establish a connection to the server.
    * `tls-timeout`: fails if worker takes too long to perform the TLS handshake with the server.
    * `resp-header-timeout`: fails if server takes too long to send headers back.

## [2019/12/11] cmaster11/overseer:1.12.4

* More reasonable release tag names.
* You can now perform [period-tests](./README.md#period-tests):

> What if you want to test how many times your web service fails in 1 minute?

## [2019/12/10] cmaster11/overseer:release-1.12

* Overseer runs now by default multiple tests in parallel (defaults to num of CPUs). This behavior is tunable with the cli flag `-parallel`:

```
overseer worker -parallel 9
```

## [2019/10/31] cmaster11/overseer:release-1.11-5

* Added support to accept multiple status codes when performing HTTP checks:

```
https://www.google.com must run http with status 200,301
```

## [2019/10/26] cmaster11/overseer:release-1.11-4

* Added support for queue-bridge filters. Destination queues (`-dest-queue=overseer.results.email`) can now be filtered
by various test result tags: `-dest-queue=overseer.results.cronjobs[type=k8s-event,target=my-namespace/Job]`. More in 
the [Kubernetes example](example-kubernetes/overseer-bridge-queue.optional.yaml) and 
[source code](bridges/queue-bridge/filter.go).

protocols/http_probe.go
// HTTP Tester
//
// The HTTP tester allows you to confirm that a remote HTTP-server is
// responding correctly.  You may test the response of a HTTP GET or
// POST request.
//
// This test is invoked via input like so:
//
//    http://example.com/ must run http
//
// By default a remote HTTP-server is considered working if it responds
// with a HTTP status-code of 200, but you can change this via:
//
//    with status 301
//
// You can also allow multiple statuses by joining them with a comma:
//
//    with status 200,429
//
// Or if you do not care about the specific status-code at all, but you
// wish to see an alert when a connection-refused/failed/timeout condition
// occurs you could say:
//
//    with status any
//
// It is also possible to regard a fetch as a failure if the response body
// does not contain a particular piece of text.  For example the following
// would be regarded as a failure if my website did not contain my name
// in the body of the response:
//
//    https://steve.fi/ must run http with content 'Steve Kemp'
//
// The following would be regarded as a failure if the response body
// DOES contain a particular piece of text:
//
// https://steve.fi/ must run http with not-content 'Steve Kemp'
//
// The 'content' setting looks for a literal match in the response-body,
// if you're looking for something more flexible you can instead test that
// the response-body matches a given regular-expression:
//
//   https://steve.fi/ must run http with pattern 'Steve\s+Kemp'
//
// The following would be regarded as a failure if the response body
// DOES match a particular pattern:
//
// https://steve.fi/ must run http with not-pattern 'Steve\s+Kemp'
//
// (The regular expression will be assumed to be multi-line, and
// will also allow newlines to be matched with ".".)
//
// If your URL requires the use of HTTP basic authentication this is
// supported by adding a username and password parameter to your test,
// for example:
//
//    https://jigsaw.w3.org/HTTP/Basic/ must run http with username 'guest' with password 'guest' with content "Your browser made it"
//
// If you need to disable failures due to expired, broken, or
// otherwise bogus SSL certificates you can do so via the tls setting:
//
//    https://expired.badssl.com/ must run http with tls insecure
//
// By default tests will fail if you're probing an SSL-site which has
// a certificate which will expire within the next 14 days.  To change
// the time-period specify it explicitly like so, if not stated the
// expiration period is assumed to be days:
//
//    # seven days
//    https://steve.fi/ must run http with expiration 7d
//
//    # 12 hours (!)
//    https://steve.fi/ must run http with expiration 12h
//
// To disable the SSL-expiration checking entirely specify "any":
//
//    https://steve.fi/ must run http with expiration any
//
// Finally if you submit a "data" argument, like in this next example
// the request made will be a HTTP POST:
//
//    https://steve.fi/Security/XSS/Tutorial/filter.cgi must run http with data "text=test%20me" with content "test me"
//
// But note that you can override the HTTP-verb via:
//
//    https://example.com/ must run http with method HEAD
//
// Combining these you can submit data with a PUT method:
//
//    https://steve.fi/Security/XSS/Tutorial/filter.cgi must run http with method PUT with data "text=test%20me" with content "test me"
//
//
// NOTE: This test deliberately does not follow redirections, to allow
// enhanced testing.
//
// To follow redirects, use the option:
//
//    with follow-redirect true <- max 10 follows (default)
//
//    with follow-redirect 20 <- max 20 follows
//

package protocols

import (
	"bytes"
	"context"
	"crypto/tls"
	"fmt"
	"io/ioutil"
	"log"
	"net"
	"net/http"
	"net/url"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/cmaster11/overseer/test"
)

// HTTPTest is our object.
type HTTPTest struct {
}

// Arguments returns the names of arguments which this protocol-test
// understands, along with corresponding regular-expressions to validate
// their values.
func (s *HTTPTest) Arguments() map[string]string {
	known := map[string]string{
		"user-agent":          ".*",
		"content":             ".*",
		"not-content":         ".*",
		"data":                ".*",
		"expiration":          "^(any|[0-9]+[hd]?)$",
		"method":              "^(GET|HEAD|POST|PUT|PATCH|DELETE)$",
		"password":            ".*",
		"pattern":             ".*",
		"not-pattern":         ".*",
		"status":              "^(any|[0-9]{3}(?:,[0-9]{3})*)$",
		"tls":                 "insecure",
		"username":            ".*",
		"connect-timeout":     `^[+]?([0-9]*(\.[0-9]*)?[a-z]+)+$`,
		"connect-retries":     `^\d+$`,
		"tls-timeout":         `^[+]?([0-9]*(\.[0-9]*)?[a-z]+)+$`,
		"resp-header-timeout": `^[+]?([0-9]*(\.[0-9]*)?[a-z]+)+$`,
		"follow-redirect":     `^true|false|(\d+)$`,
	}
	return known
}

// ShouldResolveHostname returns if this protocol requires the hostname resolution of the first test argument
func (s *HTTPTest) ShouldResolveHostname() bool {
	return true
}

// Example returns sample usage-instructions for self-documentation purposes.
func (s *HTTPTest) Example() string {
	str := `
HTTP Tester
-----------
 The HTTP tester allows you to confirm that a remote HTTP-server is
 responding correctly.  You may test the response of a HTTP GET or
 POST request.

 This test is invoked via input like so:

   http://example.com/ must run http

 By default a remote HTTP-server is considered working if it responds
 with a HTTP status-code of 200, but you can change this via:

   with status 301

 You can allow multiple statuses by joining them with a comma:

   with status 200,429

 Or if you do not care about the specific status-code at all, but you
 wish to see an alert when a connection-refused/failed/timeout condition
 occurs you could say:

   with status any

 It is also possible to regard a fetch as a failure if the response body
 does not contain a particular piece of text. For example the following
 would be regarded as a failure if my website did not contain my name
 in the body of the response:

   https://steve.fi/ must run http with content 'Steve Kemp'

 The following would be regarded as a failure if the response body 
 DOES contain a particular piece of text:

   https://steve.fi/ must run http with not-content 'Steve Kemp'

 The 'content' setting looks for a literal match in the response-body,
 if you're looking for something more flexible you can instead test that
 the response-body matches a given regular-expression:

   https://steve.fi/ must run http with pattern 'Steve\s+Kemp'

 The following would be regarded as a failure if the response body 
 DOES match a particular pattern:

   https://steve.fi/ must run http with not-pattern 'Steve\s+Kemp'

 (The regular expression will be assumed to be multi-line, and
 will also allow newlines to be matched with ".".)

 If your URL requires the use of HTTP basic authentication this is
 supported by adding a username and password parameter to your test,
 for example:

   https://jigsaw.w3.org/HTTP/Basic/ must run http with username 'guest' with password 'guest' with content "Your browser made it"

 If you need to disable failures due to expired, broken, or
 otherwise bogus SSL certificates you can do so via the tls setting:

   https://expired.badssl.com/ must run http with tls insecure

 By default tests will fail if you're probing an SSL-site which has
 a certificate which will expire within the next 14 days.  To change
 the time-period specify it explicitly like so, if not stated the
 expiration period is assumed to be days:

   # seven days
   https://steve.fi/ must run http with expiration 7d

   # 12 hours (!)
   https://steve.fi/ must run http with expiration 12h

 To disable the SSL-expiration checking entirely specify "any":

   https://steve.fi/ must run http with expiration any

 Finally if you submit a "data" argument, like in this next example
 the request made will be a HTTP POST:

   https://steve.fi/Security/XSS/Tutorial/filter.cgi must run http with data "text=test%20me" with content "test me"

 But note that you can override the HTTP-verb via:

    https://example.com/ must run http with method HEAD

 Combining these you can submit data with a PUT method:

    https://steve.fi/Security/XSS/Tutorial/filter.cgi must run http with method PUT with data "text=test%20me" with content "test me"

 Do note that the HTTP-probe never follow redirections, to allow enhanced
 testing.

 To follow redirects, use the option:
 
    with follow-redirect true <- max 10 follows (default)

    with follow-redirect 20 <- max 20 follows
`
	return str
}

// RunTest is the part of our API which is invoked to actually execute a
// HTTP-test against the given URL.
//
// For the purposes of clarity this test makes a HTTP-fetch.  The `test.Test`
// structure contains our raw test, and the `target` variable contains the
// IP address against which to make the request.
//
// So:
//
//    tst.Target => "https://steve.kemp.fi/
//
//    target => "176.9.183.100"
//
func (s *HTTPTest) RunTest(tst test.Test, target string, opts test.Options) error {

	//
	// Determine the port to connect to, initially via the protocol
	// in the string, but allow the URI to override that.
	//
	// e.g: We expect
	//
	//  http://example.com/      -> 80
	//  https://example.com/     -> 443
	//  http://example.com:8080/ -> 8080
	//
	port := "80"
	u, err := url.Parse(tst.Target)
	if err != nil {
		return err
	}
	if u.Scheme == "http" {
		port = "80"
	}
	if u.Scheme == "https" {
		port = "443"
	}
	if u.Port() != "" {
		port = u.Port()
	}

	//
	// Be clear about the IP vs. the hostname.
	//
	address := target
	target = tst.Target

	// If we're running a period test, check if we can replace any related vars
	target = strings.Replace(target, "__pt-index__", strconv.Itoa(opts.PeriodTestIndex), -1)
	target = strings.Replace(target, "__pt-time-ms__", strconv.FormatInt(opts.PeriodTestStartTime, 10), -1)

	//
	// Setup a dialer which will be dual-stack
	//
	dialer := &net.Dialer{}

	if connectTimeoutString := tst.Arguments["connect-timeout"]; connectTimeoutString != "" {
		connectTimeout, errParse := time.ParseDuration(connectTimeoutString)
		if errParse != nil {
			return errParse
		}
		dialer.Timeout = connectTimeout
	}

	maxConnectRetries := 0
	if retriesString := tst.Arguments["connect-retries"]; retriesString != "" {
		_maxDialerRetries, errParse := strconv.ParseInt(retriesString, 10, 0)
		if errParse != nil {
			return errParse
		}
		maxConnectRetries = int(_maxDialerRetries)
	}

	//
	// This is where some magic happens, we want to connect and do
	// a http check on http://example.com/, but we want to do that
	// via the IP address.
	//
	// We could do that manually by connecting to http://1.2.3.4,
	// and sending the appropriate HTTP Host: header but that risks
	// a bit of complexity with SSL in particular.
	//
	// So instead we fake the address in the dialer object, so that
	// we don't rewrite anything, don't do anything manually, and
	// instead just connect to the right IP by magic.
	//
	dial := func(ctx context.Context, network, _ string) (net.Conn, error) {
		//
		// Assume an IPv4 address by default.
		//
		addr := fmt.Sprintf("%s:%s", address, port)

		//
		// If we find a ":" we know it is an IPv6 address though
		//
		if strings.Contains(address, ":") {
			addr = fmt.Sprintf("[%s]:%s", address, port)
		}

		var conn net.Conn
		var errDial error
		for retryCount := 0; retryCount <= maxConnectRetries; retryCount++ {
			//
			// Use the replaced/updated address in our connection.
			//
			conn, errDial = dialer.DialContext(ctx, network, addr)
			// On error, if we experience a connect timeout, give it another chance if there are more retries available
			if errDial != nil {
				errNet, ok := errDial.(net.Error)
				if ok && errNet.Timeout() {
					continue
				}

				// On any other error, no retries
				break
			}
			break
		}

		return conn, errDial
	}

	//
	// Create a context which uses the dial-context
	//
	// The dial-context is where the magic happens.
	//
	tr := &http.Transport{
		DialContext: dial,
	}

	if tlsTimeoutString := tst.Arguments["tls-timeout"]; tlsTimeoutString != "" {
		tlsTimeout, errParse := time.ParseDuration(tlsTimeoutString)
		if errParse != nil {
			return errParse
		}
		tr.TLSHandshakeTimeout = tlsTimeout
	}

	if headerTimeoutString := tst.Arguments["resp-header-timeout"]; headerTimeoutString != "" {
		headerTimeout, errParse := time.ParseDuration(headerTimeoutString)
		if errParse != nil {
			return errParse
		}
		tr.ResponseHeaderTimeout = headerTimeout
	}

	//
	// If we're running insecurely then ignore SSL errors
	//
	if tst.Arguments["tls"] == "insecure" {
		tr.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
	}

	// Total request timeout
	timeout := opts.Timeout
	if tst.Timeout != nil {
		timeout = *tst.Timeout
	}

	maxFollowRedirects := 0

	argFollowRedirect := tst.Arguments["follow-redirect"]
	if parsed, errParse := strconv.ParseInt(argFollowRedirect, 10, 32); errParse == nil {
		maxFollowRedirects = int(parsed)
	} else if argFollowRedirect == "true" {
		maxFollowRedirects = 10
	}

	//
	// Create a client with a timeout, disabled redirection, and
	// the magical transport we've just created.
	//
	var netClient = &http.Client{
		Timeout:   timeout,
		Transport: tr,
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			if maxFollowRedirects > 0 {
				maxFollowRedirects--
				lastRequest := via[len(via)-1]
				log.Printf("following redirect from %s to %s", lastRequest.URL.String(), req.URL.String())
				return nil
			}
			return http.ErrUseLastResponse
		},
	}

	//
	// Now we can make a request-object
	//
	var req *http.Request

	//
	// The default method is "GET"
	//
	method := "GET"

	//
	// That can be changed
	//
	if tst.Arguments["method"] != "" {
		method = tst.Arguments["method"]
	}

	//
	// If we have no data then make a GET request
	//
	if tst.Arguments["data"] == "" {
		req, err = http.NewRequest(method, target, nil)
	} else {

		//
		// Otherwise make a HTTP POST request, with
		// the specified data.
		//
		req, err = http.NewRequest(method, target,
			bytes.NewBuffer([]byte(tst.Arguments["data"])))
	}
	if err != nil {
		return err
	}

	//
	// Are we using basic-auth?
	//
	if tst.Arguments["username"] != "" {
		req.SetBasicAuth(tst.Arguments["username"],
			tst.Arguments["password"])
	}

	//
	// Set a suitable user-agent
	//
	if tst.Arguments["user-agent"] != "" {
		req.Header.Set("User-Agent", tst.Arguments["user-agent"])
	} else {
		req.Header.Set("User-Agent", "overseer/probe")
	}

	//
	// Perform the request
	//
	response, err := netClient.Do(req)
	if err != nil {
		return err
	}

	//
	// Get the body and status-code.
	//
	defer response.Body.Close()
	body, err := ioutil.ReadAll(response.Body)
	if err != nil {
		return err
	}
	status := response.StatusCode

	//
	// The default status-code we accept as OK
	//
	var allowedStatuses []int

	//
	// Did the user want to look for a specific status-code?
	//
	if tst.Arguments["status"] != "" && tst.Arguments["status"] != "any" {

		split := strings.Split(tst.Arguments["status"], ",")
		for _, statusString := range split {
			allowedStatus, errConv := strconv.Atoi(statusString)
			if errConv != nil {
				return errConv
			}

			allowedStatuses = append(allowedStatuses, allowedStatus)
		}

	} else {

		allowedStatuses = append(allowedStatuses, http.StatusOK)

	}

	//
	// See if the status-code matched our expectation(s).
	//
	// If they mis-match that means the test failed, unless the user
	// said "with status any".
	//
	if tst.Arguments["status"] != "any" {

		found := false
		for _, allowedStatus := range allowedStatuses {
			if status == allowedStatus {
				found = true
				break
			}
		}

		if !found {
			if len(allowedStatuses) == 1 {
				return fmt.Errorf("status code was %d not %d", status, allowedStatuses[0])
			}

			return fmt.Errorf("status code was %d not one of %v", status, allowedStatuses)
		}

	}

	//
	// Is the user looking for a literal body-match?
	//
	if tst.Arguments["content"] != "" {
		if !strings.Contains(string(body), tst.Arguments["content"]) {
			return fmt.Errorf("body didn't contain '%s'", tst.Arguments["content"])
		}
	}

	//
	// Is the user NOT looking for a literal body-match?
	//
	if tst.Arguments["not-content"] != "" {
		if strings.Contains(string(body), tst.Arguments["not-content"]) {
			return fmt.Errorf("body contains '%s'", tst.Arguments["not-content"])
		}
	}

	//
	// Is the user expecting a regular expression to match the content?
	//
	if tst.Arguments["pattern"] != "" {
		var re *regexp.Regexp
		re, err = regexp.Compile("(?ms)" + tst.Arguments["pattern"])
		if err != nil {
			return err
		}

		// Skip unless this handler matches the filter.
		match := re.FindAllStringSubmatch(string(body), -1)
		if len(match) < 1 {
			return fmt.Errorf("body didn't match the regular expression '%s'", tst.Arguments["pattern"])
		}
	}

	//
	// Is the user NOT expecting a regular expression to match the content?
	//
	if tst.Arguments["not-pattern"] != "" {
		var re *regexp.Regexp
		re, err = regexp.Compile("(?ms)" + tst.Arguments["not-pattern"])
		if err != nil {
			return err
		}

		// Skip unless this handler matches the filter.
		match := re.FindAllStringSubmatch(string(body), -1)
		if len(match) > 0 {
			return fmt.Errorf("body matched the regular expression '%s'", tst.Arguments["not-pattern"])
		}
	}

	//
	// If we reached here then our actual test was fine.
	//
	// However as a special extension we're going to test the
	// certificate expiration date for any SSL sites.  We'll
	// do that now.
	//
	if strings.HasPrefix(tst.Target, "https:") {

		//
		// The default expiration-time 14 days.
		//
		period := 14 * 24

		//
		// If the validity was set to `any` that means we just
		// don't care, so we don't even need to test the result.
		//
		if tst.Arguments["expiration"] == "any" {
			return nil
		}

		//
		// The user might have specified a different period
		// in hours / days.
		//
		expire := tst.Arguments["expiration"]
		if expire != "" {

			//
			// How much to scale the given figure by
			//
			// By default if no units are specified we'll
			// assume the figure is in days, so no scaling
			// is required.
			//
			mul := 1

			// Days?
			if strings.HasSuffix(expire, "d") {
				expire = strings.Replace(expire, "d", "", -1)
				mul = 24
			}

			// Hours?
			if strings.HasSuffix(expire, "h") {
				expire = strings.Replace(expire, "h", "", -1)
				mul = 1
			}

			// Get the period.
			period, err = strconv.Atoi(expire)
			if err != nil {
				return err
			}

			//
			// Multiply by our multiplier.
			//
			period *= mul
		}

		//
		// Check the expiration
		//
		hours, cn, errExpire := s.SSLExpiration(tst.Target, opts.Verbose)
		if errExpire == nil {
			// Is the age too short?
			if int64(hours) < int64(period) {

				return fmt.Errorf("SSL certificate '%s' will expire in %d hours (%d days)", cn, hours, int(hours/24))
			}
		}

	}

	//
	// If we reached here all is OK
	//
	return nil
}

// SSLExpiration returns the number of hours remaining for a given
// SSL certificate chain.
func (s *HTTPTest) SSLExpiration(host string, verbose bool) (int64, string, error) {

	// Expiry time, in hours
	var hours int64
	hours = -1

	// The common-name of the certificate involved.
	cn := ""

	//
	// If the string matches https://, then strip it off
	//
	re, err := regexp.Compile(`^https:\/\/([^\/]+)`)
	if err != nil {
		return 0, "", err
	}
	res := re.FindAllStringSubmatch(host, -1)
	for _, v := range res {
		host = v[1]
	}

	//
	// If no port is specified default to :443
	//
	p := strings.Index(host, ":")
	if p == -1 {
		host += ":443"
	}

	//
	// Show what we're doing.
	//
	if verbose {
		fmt.Printf("SSLExpiration testing: %s\n", host)
	}

	conn, err := tls.Dial("tcp", host, nil)
	if err != nil {
		return 0, "", err
	}
	defer conn.Close()

	timeNow := time.Now()
	for _, chain := range conn.ConnectionState().VerifiedChains {
		for _, cert := range chain {

			// Get the expiration time, in hours.
			expiresIn := int64(cert.NotAfter.Sub(timeNow).Hours())

			if verbose {
				fmt.Printf("SSLExpiration - certificate: %s expires in %d hours (%d days)\n", cert.Subject.CommonName, expiresIn, expiresIn/24)
			}

			// If we've not checked anything this is the benchmark
			if hours == -1 {
				hours = expiresIn
				cn = cert.Subject.CommonName
			} else {
				// Otherwise replace our result if the
				// certificate is going to expire more
				// recently than the current "winner".
				if expiresIn < hours {
					hours = expiresIn
					cn = cert.Subject.CommonName
				}
			}
		}
	}

	return hours, cn, nil
}

func (s *HTTPTest) GetUniqueHashForTest(tst test.Test, opts test.Options) *string {
	return nil
}

// init is used to dynamically register our protocol-tester.
func init() {
	Register("http", func() ProtocolTest {
		return &HTTPTest{}
	})
}
